PFW-1312 Define layout of LCD for MMU error screens

gudnimg opened the pull request:
See JIRA ticket: https://dev.prusa3d.com/browse/PFW-1312

**Currently the PR is a Draft and is a work in progress!**

TODO:

- [x] I have created an inconsistency in the menus which have choices. Previously the left most choice was 1, and Middle choice was 0. In my changes it is: Left most is 0, Middle is 1 and right most is 2.
- [x] Do we want to return to the status screen after running a MMU command in the menus? => We want to return to the last menu. Example: When running **Load filament** -> **Filament 1** we would like to return back to the **Load filament** menu.
- [x] `third_col` might not be needed and could be constant. When three choices are used we don't really have much freedom where to place the third column/choice.
- [x] Turn `BUTTON_OP_HIGH_NIBBLE_MSK` and `BUTTON_OP_LOW_NIBBLE_MSK` into macros such that `BUTTON_OP_HIGH_NIBBLE_MSK` is `BUTTON_OP_HIGH_NIBBLE(X) ( (X & 0xF0) >> 4 )`
- [x] Set the middle button as the default choice when there are three choices. Else the left most button when there are two choices. This requires changing the parameter type for `default_first` from `bool` to `uint8_t` in `lcd_show_multiscreen_message_with_choices_and_wait_P`.
- [x] Create an `enum` for the button choices to make the code more readable. `enum { LEFT_BUTTON_CHOICE = 0, MIDDLE_BUTTON_CHOICE = 1, RIGHT_BUTTON_CHOICE = 2 };`
- [x] Change name of variable `yes` in `lcd_show_multiscreen_message_with_choices_and_wait_P` to something more descriptive. (it no longer represents only yes or no)

----

List of changes:

----

Change in memory consumption:

----

wavexx committed:
Unify fullscreen/multiscreen/yes-no prompt message handling

Modify lcd_show_multiscreen_message_two_choices_and_wait_P to also
handle single-screen or empty (no-clear) prompts, making other functions
redundant. Saves 76 bytes.

Change existing functions to simply call
lcd_show_multiscreen_message_two_choices_and_wait_P with the correct
arguments.

This changes the prompt of existing Yes/No messages: the previous prompt
would use the last two lines of the LCD, while the new prompt is using
just the last line of the LCD instead.

Translation do not require updates, since the Yes/No translation was
already the same in both implementations.

wavexx committed:
Allow choice positioning. Balance Yes/No messages.

Add an additional parameter to control the position of second choice
prompt position (while defaulting to the old).

This allows Yes/No prompts to be equally spaced.

gudnimg committed:
Merge pull request #1 from wavexx/unify_fullscreen_msg_handling

Merge PR-3090

gudnimg committed:
add that downwards facing >> character

gudnimg committed:
Print the MMU progress into status screen

gudnimg committed:
Remove class LcdUpdateDisabler

Calling lcd_update_enable directly gives us more control

This also save a bit of flash memory

gudnimg committed:
Allow LCD to be updated in manage_response loop

Also removed one TODO which David mentioned is not needed anymore

gudnimg committed:
Update lcd_show_multiscreen_message_two_choices_and_wait_P

to support 3 choices

gudnimg committed:
Don't disable LCD updates in menu_item_function_P

gudnimg committed:
Implement PFW-1312 in MMU code

gudnimg committed:
Checkpoint:

- Rename lcd_show_multiscreen_message_two_choices_and_wait_P to lcd_show_multiscreen_message_with_choices_and_wait_P

- Move MMU error screen from ReportError() to ReportErrorHook()

- Fix the menu selection to work for menus with either two choices or three choices

- The buttons shown on the MMU error menu are now determined by the definition of btnOperation[]

gudnimg committed:
Remove unused includes after I moved the MMU error screen

DRracer commented:
@gudnimg 
- order of buttons - actually we may prefer having the middle button as the default one - that may require reordering the button definition, but nothing more. If other dialogs behave like this, let's do it that way.
- return to status screen after running a command from menu - not necessarily, it would be nice to stay in the menu tree if possible
- third column/button/action will be removed

davidkerny commented:
When running the MMU commands in the menu, returning to the last menu would be the best option. 
For instance, if you hit **Load filament > Filament 1** - after a successful load, you'd be back in the filament positions list so you can load another filament.

gudnimg commented:
I dropped the commit where I was returning to the status screen after loading filament. Will make sure the behaviour is as you have both described @davidkerny @DRracer 

I'm thinking I should drop this commit also for the same reason: https://github.com/DRracer/Prusa-Firmware/pull/15/commits/c2b55b994e72b32243d563c264f564c89e42d693 as it is also not technically related to PFW-1312. This will prevent the user from using MK3S+ while the MMU is running a command. (Edit: will need to test this though as I don't recall if this affects the error screen)

gudnimg commented:
> third column/button/action will be removed

@DRracer I'm not sure what you mean here. Its my understanding that the '`More`' button should always be present. That is how I understood the following (bold) text in the JIRA ticket:

> Button "More" allows displaying of the additional text on next pages **_and will always be present on the screen_** (we can tweak the error pages' button definition if necessary).
"W" will be the double arrow used to suggest more text is available upon clicking the knob.

DRracer commented:
@gudnimg 
> @DRracer I'm not sure what you mean here. Its my understanding that the '`More`' button should always be present. That is how I understood the following (bold) text in the JIRA ticket:

For FINDA and FSensor not triggering there are 3 actions in the Prusa-Error-Codes definition. One of them is "Slow", which both doesn't work and is not used - this button will get removed from the definition and thus all error screens will only have 2 buttons, which fits onto our LCD. Yes, "More" will be always present.

gudnimg committed:
Add helper macros to parse nibbles

Also made the button operations and nibbles constant since they should not be modified.

gudnimg committed:
Checkpoint

- combined lcd_show_three_choices_prompt_P and lcd_show_two_choices_prompt_P into one function called lcd_show_choices_prompt_P
- Removed third_col parameter
- Added an enum to make the button selection result more readable

gudnimg committed:
Make all choice menus return consistent values

Such that left = 0, middle = 1, and right = 2

gudnimg committed:
Fixup

gudnimg committed:
Merge branch 'mk3-mmu' into PFW-1312

gudnimg committed:
Update the error menu after merging new changes

Will test this tomorrow

gudnimg committed:
Change name of variable 'yes' to be more descriptive

gudnimg committed:
Change default selection parameter from bool to uint8_t

This allows us make any of the three button choices the default selection

Commit:
Fix compile-time conversion MMU2 ErrorCode -> Prusa-Error-Code

and hide some of the implementation details into mmu2_error_converter.cpp
which makes the code in mmu_reporting.cpp much easier to read.

Commit:
Make sure the MMU error title never overflows on the LCD

Commit:
Fix FW update needed button order
